rfcs/text/refactoring-date-component.md

```diff
@@ -0,0 +1,330 @@
+    const weekdayLong = localize.day(weekday);     
+    const weekdayShort = localize.day(weekday, { width: "abbreviated" });
+    
+    return (
+      <Weekday className={className} title={weekdayLong}>
+        <abbr title={weekdayLong}>{weekdayShort}</abbr>
+      </Weekday>
+    );
+  }}
+```
+
+As with the existing component we will store a `lastValidValue` in internal state and use that to control the date displayed in the `DatePicker`. This can be achieved by leveraging a `useEffect` hook to update the state value if a new `inputValue` parses to a valid Date. We will also keep the same implementation the current component has with regards to rendering the `DatePicker` in a `portal` using the existing `Popover` component.
+
+```jsx
+  // date/__internal__/date-picker/date-picker.component.js
+  import {
+    localeMap,
+    isValidDate,
+    parseToDate
+  } from "./__internal__/date-utils";
+
+  const Picker = ({
+    inputValue,
+    selectedDate,
+    onDayClick,
+    isOpen,
+    disablePortal
+  }) => {
+    const [lastValidValue, setLastValidValue] = useState(inputValue);
+    const { locale, date } = useContext(LocaleContext);
+    const { formats } = date;
+    const { localize, options } = localeMap[locale()];
+    const format = formats.javascript();
+    const { weekStartsOn } = options;
+
+    if (!isOpen) return null;
+
+    useEffect(() => {
+      // check if inputValue parses to valid date and update lastValidValue string if true
+    }, [
+      format,
+      inputValue,
+      lastValidValue,
+    ]);
+
+    return (
+      <Popover
+        ...
+      >
+      <StyledDayPicker>
+        <DayPicker
+          firstDayOfWeek={weekStartsOn}
+          onDayClick={onDayClick}
+          selectedDays={selectedDate}
+          month={ parseToDate(lastValidValue, format) }
+          weekdayElement={renderWeekdayElement}
+          ...
+        />
+      </StyledDayPicker>
+    </Popover>
+    );
+  };
+```
+
+# Drawbacks
+
+Carrying out this proposal will mean introducing several breaking changes. Writing the new component as a functional component will mean that any consuming project that has extended the previous class-based component will now need to update to use composition instead. We will also be removing the utils we currently export and any new ones we write will be in an `__internal__` directory. By moving the `DatePicker`, `Weekday` and `Navbar` components to an `__internal__` directory we will also be making a breaking change, but this will be for the better in the long run as these components should not be directly imported by consuming projects and we can then make changes to their interfaces without needing to have them be breaking. The `Date` component is used extensively by our consumers so these breaking changes are likely to have widespread impact. Although it is certainly arguable that the current implementation has enough issues with it that this too is having significant impact.
```

Comment:
Should we stage the breaking changes? E.g. remove the shared utils, make the components `__internal__` then rewrite date?

Reply:
Good idea, I think we could do this as two breaking changes, move the utils and components to `__internal__` and then carry out the rewrite of `Date`.

Reply:
@chrisbarber86 I'd missed that we're actually already moving all our utils to an `__internal__` directory as part of separate ticket so the work will be carried out as two breaking changes, I'll update the document to reflect this